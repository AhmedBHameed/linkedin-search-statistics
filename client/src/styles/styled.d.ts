import 'styled-components';

/**
 * For more information whey we need this file @see https://styled-components.com/docs/api#create-a-declarations-file
 */
declare module 'styled-components' {
  export interface DefaultTheme {
    colors: {
      blue5: string;
      blue50: string;
      blue90: string;
      red: string;
      highlightText: string;
    };
  }
}
