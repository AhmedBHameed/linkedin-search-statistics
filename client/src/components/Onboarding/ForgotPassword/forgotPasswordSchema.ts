import {Joi, requiredEmail} from '../../../util/validations';
import {ForgotPasswordModel} from './models/ForgotPasswordModel';

const forgotPasswordSchema = Joi.object<ForgotPasswordModel>({
  email: requiredEmail(),
});

export default forgotPasswordSchema;
