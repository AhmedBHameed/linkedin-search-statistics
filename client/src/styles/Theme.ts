/**
 * The number 50 is the scale of the color according to Figma
 * @see https://www.figma.com/file/WCipxn54SiYJSVpZ3uTUW5/BiomeDx-Styleguide-und-Report-%234?node-id=13%3A140
 */

const highlightText = '#ffc069';

const lightTheme = {
  red: 'red',
  blue5: '#e7f2fe',
  blue50: '#047CB3',
  blue90: '#011e2f',
  highlightText,
};

const darkTheme = {
  red: '#ff7777',
  blue5: '#e7f2fe',
  blue50: '#047CB3',
  blue90: '#011e2f',
  highlightText,
};

export type ThemeProps = typeof darkTheme;
export {lightTheme, darkTheme};
