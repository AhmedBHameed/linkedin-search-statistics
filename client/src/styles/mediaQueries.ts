export const sm = 576;
export const md = 768;
export const lg = 992;
export const xl = 1200;
export const xxl = 1600;

const devices = {
  smMediaQuery: `(min-width: ${sm}px)`,
  mdMediaQuery: `(min-width: ${md}px)`,
  lgMediaQuery: `(min-width: ${lg}px)`,
  xlMediaQuery: `(min-width: ${xl}px)`,
  xxlMediaQuery: `(min-width: ${xxl}px)`,
};

export default devices;
