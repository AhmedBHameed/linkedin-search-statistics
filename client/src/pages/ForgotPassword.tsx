import {MailOutlined} from '@ant-design/icons';
import {joiResolver} from '@hookform/resolvers/joi';
import {Card, Col, Row} from 'antd';
import React, {useCallback, useState} from 'react';
import {Controller, useForm} from 'react-hook-form';
import {Link} from 'react-router-dom';
import styled from 'styled-components';
import LoadingButton from '../components/Buttons/LoadingButton';

import {FormControl, InputField} from '../components/Forms/Form';
import forgotPasswordSchema from '../components/Onboarding/ForgotPassword/forgotPasswordSchema';
import {ForgotPasswordModel} from '../components/Onboarding/ForgotPassword/models/ForgotPasswordModel';
import auth from '../components/Onboarding/util/auth';
import {
  FormTitle,
  OnboardingForm,
  StyledBrand,
  StyledRow,
} from '../components/shared/onboardingStyles';
import useToastMessage from '../components/ToastMessage/hooks/toastMessageHook';
import NAVIGATION_ROUTES from '../config/NavigationRoutes';
import {callTryCatch} from '../util/callTryCatch';

// const RedirectToSignupButton = styled(Button)`
//   padding: 0;
// `;

const FullWidthLoadingButton = styled(LoadingButton)`
  width: 100%;
  margin: 2rem 0;
`;

const ForgotPassword: React.FC = () => {
  const [loading, setLoading] = useState(false);
  const {notifyError, notifySuccess} = useToastMessage();

  const {
    control,
    formState: {errors},
    handleSubmit,
  } = useForm<ForgotPasswordModel>({
    resolver: joiResolver(forgotPasswordSchema),
    defaultValues: {
      email: '',
    },
  });

  const forgotPassword = useCallback(
    async (formData: ForgotPasswordModel) => {
      setLoading(true);
      const [error] = await callTryCatch(async () =>
        auth.forgotPassword({
          email: formData.email,
        })
      );

      setLoading(false);

      if (error) {
        // eslint-disable-next-line no-console
        console.error('<Signup />', error);
        notifyError({
          message: 'Forgot password',
          description: 'Email not found in our database',
        });

        return;
      }

      notifySuccess({
        message: 'Forgot password',
        description: 'We will help you retrieve you password.',
      });
    },
    [notifyError, notifySuccess]
  );

  const emailError = errors.email?.message;

  return (
    <StyledRow align="middle" justify="center">
      <Col lg={8} md={10} sm={12} xs={24}>
        <OnboardingForm
          onSubmit={handleSubmit(forgotPassword)}
          testId="forgot-password-form"
        >
          <Card>
            <Row align="middle" justify="center">
              <StyledBrand />
            </Row>

            <Row>
              <Col xs={24}>
                <FormTitle level={5}>Forgot password</FormTitle>
              </Col>
            </Row>

            <Row>
              <Col xs={24}>
                <FormControl error={emailError} label="Email">
                  <Controller
                    control={control}
                    name="email"
                    render={({field: {onChange, value}}) => (
                      <InputField
                        ariaLabel="email"
                        error={!!emailError}
                        name="email"
                        onChange={onChange}
                        placeholder="Email"
                        prefixIcon={<MailOutlined />}
                        size="large"
                        testId="email-input"
                        value={value}
                      />
                    )}
                  />
                </FormControl>
              </Col>
            </Row>

            <Row justify="end">
              <Link to={NAVIGATION_ROUTES.login.path}>
                Return back to login.
              </Link>
            </Row>

            <Row>
              <FullWidthLoadingButton
                disabled={loading}
                htmlType="submit"
                icon={<MailOutlined />}
                loading={loading}
                size="large"
                testId="forgot-password-submit-button"
              >
                Submit
              </FullWidthLoadingButton>
            </Row>

            <Row align="middle" justify="center">
              <Link to={NAVIGATION_ROUTES.signup.path}>
                Don&apos;t have account? register here.
              </Link>
            </Row>
          </Card>
        </OnboardingForm>
      </Col>
    </StyledRow>
  );
};

export default ForgotPassword;
