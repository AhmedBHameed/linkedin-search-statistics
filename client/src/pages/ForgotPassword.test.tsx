import {act, fireEvent, render, waitFor} from '@testing-library/react';
import {Space} from 'antd';

import {BrowserRouter as Router} from 'react-router-dom';
import {ThemeProvider} from 'styled-components';

import {mockConsoleErrorFun, mockMediaQuery} from '../setupTests';
import {md} from '../styles/mediaQueries';
import {lightTheme} from '../styles/Theme';
import {mockForgotPassword} from '../test/utils/generate';
import ForgotPassword from './ForgotPassword';

mockConsoleErrorFun();

const mockNotifyError = jest.fn();
const mockNotifySuccess = jest.fn();
const mockCloseNotification = jest.fn();
jest.mock('../components/ToastMessage/hooks/toastMessageHook', () => () => ({
  notifyError: mockNotifyError,
  notifySuccess: mockNotifySuccess,
  closeNotification: mockCloseNotification,
}));

const mockGoToSignup = jest.fn();
jest.mock('../components/Signup/hooks/navigateToSignupHook', () => () => ({
  goToSignup: mockGoToSignup,
}));

const mockForgotPasswordHook = jest.fn();
jest.mock(
  '../components/ForgotPassword/hooks/forgotPasswordHook',
  () => () => ({
    handleForgotPassword: mockForgotPasswordHook,
  })
);

const MOCK_FORGOT_PASSWORD = mockForgotPassword();

function renderComponent() {
  mockMediaQuery(md, md);

  return render(
    <ThemeProvider theme={{colors: lightTheme}}>
      <Router>
        <Space>
          <ForgotPassword />
        </Space>
      </Router>
    </ThemeProvider>
  );
}

describe('ForgotPassword', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  it('matches snapshot', () => {
    const {container} = renderComponent();

    expect(container).toMatchSnapshot();
  });

  describe('Form validation', () => {
    it('shows helper text as a hint when submitting invalid data.', async () => {
      const {findAllByTestId, getByTestId} = renderComponent();

      act(() => {
        fireEvent.submit(getByTestId('forgot-password-submit-button'));
      });

      const inputErrors = await findAllByTestId('input-error');

      expect(inputErrors[0].textContent).toBe('Field required');
      expect(mockGoToSignup).toHaveBeenCalledTimes(0);
      expect(mockNotifyError).toHaveBeenCalledTimes(0);
      expect(mockForgotPasswordHook).toHaveBeenCalledTimes(0);
    });

    it('submits valid data and show successful toast message.', async () => {
      const {getByTestId} = renderComponent();

      await act(async () => {
        fireEvent.change(getByTestId('email-input'), {
          target: {value: MOCK_FORGOT_PASSWORD.email},
        });
        await waitFor(() => {});

        fireEvent.submit(getByTestId('forgot-password-form'));
        await waitFor(() => {});

        jest.advanceTimersByTime(3000);
      });

      expect(mockForgotPasswordHook).toHaveBeenCalledTimes(1);
      expect(mockForgotPasswordHook).toHaveBeenCalledWith(MOCK_FORGOT_PASSWORD);

      expect(mockGoToSignup).toHaveBeenCalledTimes(0);
      expect(mockNotifyError).toHaveBeenCalledTimes(0);

      expect(mockNotifySuccess).toHaveBeenCalledTimes(1);
      expect(mockNotifySuccess).toHaveBeenCalledWith({
        message: 'forgotPassword.success.forgotPasswordSubmitted.title',
        description: 'forgotPassword.success.forgotPasswordSubmitted.message',
      });
    });

    it('triggers toast message when async throw an error', async () => {
      mockForgotPasswordHook.mockRejectedValueOnce(new Error('async error'));

      const {getByTestId} = renderComponent();

      await act(async () => {
        fireEvent.change(getByTestId('email-input'), {
          target: {value: MOCK_FORGOT_PASSWORD.email},
        });
        await waitFor(() => {});

        fireEvent.submit(getByTestId('forgot-password-form'));
        await waitFor(() => {});

        jest.advanceTimersByTime(3000);
      });

      expect(mockForgotPasswordHook).toHaveBeenCalledTimes(1);
      expect(mockForgotPasswordHook).toHaveBeenCalledWith(MOCK_FORGOT_PASSWORD);
      // close-notification-button
      expect(mockNotifySuccess).toHaveBeenCalledTimes(0);

      expect(mockGoToSignup).toHaveBeenCalledTimes(0);

      expect(mockNotifyError).toHaveBeenCalledTimes(1);
    });
  });
});
