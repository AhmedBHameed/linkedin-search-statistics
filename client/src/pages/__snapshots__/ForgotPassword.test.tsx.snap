// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword matches snapshot 1`] = `
<div>
  <div
    class="ant-space ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
    >
      <div
        class="ant-row ant-row-center ant-row-middle sc-pNWdM oTyik"
      >
        <div
          class="ant-col ant-col-xs-24 ant-col-sm-12 ant-col-md-10 ant-col-lg-8"
        >
          <form
            class="sc-jrsJWt fvCjFZ"
            data-testid="forgot-password-form"
          >
            <div
              class="ant-card ant-card-bordered"
            >
              <div
                class="ant-card-body"
              >
                <div
                  class="ant-row ant-row-center ant-row-middle"
                >
                  <svg
                    class="sc-bdnxRM goXlsq"
                  >
                    brand-light-theme.svg
                  </svg>
                </div>
                <div
                  class="ant-row"
                >
                  <div
                    class="ant-col ant-col-xs-24"
                  >
                    <h5
                      class="ant-typography sc-kEqXSa dYmnsH"
                    >
                      forgotPassword.description
                    </h5>
                  </div>
                </div>
                <div
                  class="ant-row"
                >
                  <div
                    class="ant-col ant-col-xs-24"
                  >
                    <div
                      class="ant-row ant-form-item sc-jSFjdj byNJCq"
                    >
                      <div
                        class="ant-col ant-form-item-label"
                      >
                        <label
                          class=""
                          title="common.emailLabel"
                        >
                          common.emailLabel
                        </label>
                      </div>
                      <div
                        class="ant-col ant-form-item-control"
                      >
                        <div
                          class="ant-form-item-control-input"
                        >
                          <div
                            class="ant-form-item-control-input-content"
                          >
                            <span
                              class="ant-input-affix-wrapper ant-input-affix-wrapper-lg"
                            >
                              <span
                                class="ant-input-prefix"
                              >
                                <span
                                  aria-label="mail"
                                  class="anticon anticon-mail"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="mail"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="64 64 896 896"
                                    width="1em"
                                  >
                                    <path
                                      d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 110.8V792H136V270.8l-27.6-21.5 39.3-50.5 42.8 33.3h643.1l42.8-33.3 39.3 50.5-27.7 21.5zM833.6 232L512 482 190.4 232l-42.8-33.3-39.3 50.5 27.6 21.5 341.6 265.6a55.99 55.99 0 0068.7 0L888 270.8l27.6-21.5-39.3-50.5-42.7 33.2z"
                                    />
                                  </svg>
                                </span>
                              </span>
                              <input
                                aria-label="email"
                                class="ant-input ant-input-lg"
                                data-testid="email-input"
                                name="email"
                                placeholder="common.emailPlaceholder"
                                type="text"
                                value=""
                              />
                            </span>
                            <small
                              class="sc-gKAaRy kDbdUP ant-form-item-explain ant-form-item-explain-error"
                              data-testid=""
                            >
                              Â 
                            </small>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-row ant-row-end"
                >
                  <a
                    href="/login"
                  >
                    forgotPassword.redirectToLogin
                  </a>
                </div>
                <div
                  class="ant-row"
                >
                  <button
                    class="ant-btn ant-btn-primary ant-btn-lg sc-dIsUp izwWel"
                    data-testid="forgot-password-submit-button"
                    type="submit"
                  >
                    <span
                      aria-label="mail"
                      class="anticon anticon-mail"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="mail"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 110.8V792H136V270.8l-27.6-21.5 39.3-50.5 42.8 33.3h643.1l42.8-33.3 39.3 50.5-27.7 21.5zM833.6 232L512 482 190.4 232l-42.8-33.3-39.3 50.5 27.6 21.5 341.6 265.6a55.99 55.99 0 0068.7 0L888 270.8l27.6-21.5-39.3-50.5-42.7 33.2z"
                        />
                      </svg>
                    </span>
                    <span
                      class="sc-dlnjwi jsqGlf"
                    >
                      forgotPassword.actionButton
                    </span>
                  </button>
                </div>
                <div
                  class="ant-row ant-row-center ant-row-middle"
                >
                  <a
                    href="/signup"
                  >
                    common.notRegistered
                  </a>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
